lib/kube/proxy/server.go

```diff
@@ -192,8 +173,22 @@ func (t *TLSServer) Serve(listener net.Listener) error {
 	}
 	t.mu.Unlock()
 
-	if t.heartbeat != nil {
-		go t.heartbeat.Run()
+	// Start the heartbeat to announce kubernetes_service presence.
+	//
+	// Only announce when running in an actual kube_server, or when
+	// running in proxy_service with local kube credentials. This means that
+	// proxy_service will pretend to also be kube_server.
+	if t.KubeServiceType == KubeService ||
+		(t.KubeServiceType == LegacyProxyService && len(t.fwd.kubeClusters()) > 0) {
+		log.Debugf("Starting kubernetes_service heartbeats for %q", t.Component)
+		for _, cluster := range t.fwd.kubeClusters() {
+
+			if err := t.startHeartbeat(t.closeContext, cluster); err != nil {
```

Comment:
If each kubernetes cluster is treated as independent resource should we also start and handle dynamic labels ?

Reply:
AFAIK dynamic labels are independent of the attached resource (maybe I am wrong??).
Given that, and since dynamic labels, for now, are provided at the server config level (you cannot specify them in `kubeconfig`), so they should be equal for any cluster specified there.

Reply:
All right, the kube cluster resource inherits the dynamic label defined for  kube agent in the `kubernetes_services` so it should be fine if right now we cannot  set dynamic label per cluster resource.